src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts

```diff
@@ -441,66 +441,81 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
 		// A save participant can still change the model now and since we are so close to saving
 		// we do not want to trigger another auto save or similar, so we block this
 		// In addition we update our version right after in case it changed because of a model change
+		let saveParticipantPromise = TPromise.as(versionId);
+
 		if (TextFileEditorModel.saveParticipant) {
-			this.blockModelContentChange = true;
-			try {
-				TextFileEditorModel.saveParticipant.participate(this, { isAutoSaved });
-			} finally {
+			saveParticipantPromise = TPromise.as(undefined).then(() => {
+				this.blockModelContentChange = true;
+				return TextFileEditorModel.saveParticipant.participate(this, { isAutoSaved });
+			}).then(() => {
 				this.blockModelContentChange = false;
-			}
-			versionId = this.versionId;
+				return this.versionId;
+			}, err => {
+				this.blockModelContentChange = false;
+				return TPromise.wrapError(err);
+			});
 		}
 
-		// Clear error flag since we are trying to save again
-		this.inErrorMode = false;
+		this.mapPendingSaveToVersionId[versionId] = saveParticipantPromise.then(newVersionId => {
```

Comment:
@jrieken is there a reason why you chose to store the save participant into `mapPendingSaveToVersionId`, then delete it and later store it in `mapPendingSaveToVersionId` again for the `updateContent` call?

Easier to read and understand, I suggest to store the save participant promise into `mapPendingSaveToVersionId` as you do now, but do not delete it right after. Rather let the promise from `this.fileService.updateContent` also contribute to the overall pending save. 

In other words, a pending save is the sum of running save participants and saving to disk.